router/router_test.go

```diff
@@ -108,3 +104,129 @@ func TestRingpopRouterMakeRemoteClientCached(t *testing.T) {
 }
+
+func TestRingpopRouterRemoveExistingClient(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.removeClient(hostport)
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveNonExistingClient(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	assert.Equal(t, nil, router.clientCache[hostport])
+	router.removeClient(hostport)
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveClientOnSwimFaultyEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Faulty,
+			},
+		},
+	})
+
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterRemoveClientOnSwimLeaveEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Leave,
+			},
+		},
+	})
+
+	assert.Equal(t, nil, router.clientCache[hostport])
+}
+
+func TestRingpopRouterNotRemoveClientOnSwimSuspectEvent(t *testing.T) {
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+
+	router := New(rp, nil, nil).(*router)
+
+	hostport := "127.0.0.1:3000"
+	router.clientCache[hostport] = "Some client"
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+
+	router.HandleEvent(swim.MemberlistChangesReceivedEvent{
+		Changes: []swim.Change{
+			swim.Change{
+				Address: hostport,
+				Status:  swim.Suspect,
+			},
+		},
+	})
+
+	assert.Equal(t, "Some client", router.clientCache[hostport])
+}
+
+func TestRingpopRouterGetClientForwardLookupError(t *testing.T) {
+	cf := &mocks.ClientFactory{}
+	cf.On("GetLocalClient").Return(nil)
+
+	stubError := errors.New("Ringpop not started")
+
+	rp := &mocks.Ringpop{}
+	rp.On("RegisterListener", mock.Anything).Return()
+	rp.On("Lookup", "hello").Return("", stubError)
+	rp.On("WhoAmI").Return("127.0.0.1:3000", nil)
+
+	router := New(rp, cf, nil)
+
+	_, err := router.GetClient("hello")
+	assert.Equal(t, errors.New("Ringpop not started"), err)
```

Comment:
"not started" should is inconsistent with the other ways we talk about Ringpop. "not ready" is more consistent.